src/kernel/cuda/spmm.cuh

```diff
@@ -0,0 +1,272 @@
+
+template <typename T>
+__device__ __forceinline__ T _ldg(T* addr) {
+#if __CUDA_ARCH__ >= 350
+  return __ldg(addr);
+#else
+  return *addr;
+#endif
+}
+
+template <typename Idx, typename DType,
+          typename BinaryOp, typename ReduceOp,
+          bool UseBcast = false, bool UseIdx = false>
+__global__ void SpMMCooKernel(
+  const DType *ufeat, const DType *efeat, DType *out, Idx *arg_u, Idx *arg_e,
+  const Idx *row, const Idx *col, const Idx* edge_map,
+  int64_t N, int64_t M, int64_t E,
+  int64_t *ubcast_off, int64_t *ebcast_off,
+  int64_t ufeat_len, int64_t efeat_len, int64_t out_len) {
+  // SPMM with COO.
+  Idx ty = blockIdx.y * blockDim.y + threadIdx.y;
+  const Idx stride_y = blockDim.y * gridDim.y;
+  while (ty < E) {
+    const Idx src = _ldg(row + ty);
+    const Idx dst = _ldg(col + ty);
+    const Idx eid = UseIdx ? _ldg(edge_map + ty) : ty;
+    int64_t tx = blockIdx.x * blockDim.x + threadIdx.x;
+    const int64_t stride_x = blockDim.x * gridDim.x;
+    const DType* uoff = BinaryOp::use_lhs ? (ufeat + src * ufeat_len): nullptr;
+    const DType* eoff = BinaryOp::use_rhs ? (efeat + eid * efeat_len): nullptr;
+    DType* outoff = out + dst * out_len;
+    while (tx < out_len) {
+      const int64_t lhs_add = UseBcast ? ubcast_off[tx] : tx;
+      const int64_t rhs_add = UseBcast ? ebcast_off[tx] : tx;
+      DType val = BinaryOp::Call(uoff + lhs_add, eoff + rhs_add);
+      Idx* arguoff = (ReduceOp::require_arg && BinaryOp::use_lhs) ?
+        (arg_u + dst * out_len + tx): nullptr;
+      Idx* argeoff = (ReduceOp::require_arg && BinaryOp::use_rhs) ?
+        (arg_e + dst * out_len + tx): nullptr;
+      ReduceOp::Call(outoff + tx, arguoff, argeoff, val, src, eid);
+      tx += stride_x;
+    }
+    ty += stride_y;
+  }
+}
+
+template <typename Idx, typename DType,
+          typename BinaryOp, typename ReduceOp,
+          bool UseBcast = false, bool UseIdx = false>
+__global__ void ArgSpMMCooKernel(
+  const DType *ufeat, const DType *efeat, DType *out, Idx *arg_u, Idx *arg_e,
+  const Idx *row, const Idx *col, const Idx* edge_map,
+  int64_t N, int64_t M, int64_t E,
+  int64_t *ubcast_off, int64_t *ebcast_off,
+  int64_t ufeat_len, int64_t efeat_len, int64_t out_len) {
+  // SPMM with COO arg max/min.
+  Idx ty = blockIdx.y * blockDim.y + threadIdx.y;
+  const Idx stride_y = blockDim.y * gridDim.y;
+  while (ty < E) {
+    const Idx src = _ldg(row + ty);
+    const Idx dst = _ldg(col + ty);
+    const Idx eid = UseIdx ? _ldg(edge_map + ty) : ty;
+    int64_t tx = blockIdx.x * blockDim.x + threadIdx.x;
+    const int64_t stride_x = blockDim.x * gridDim.x;
+    const DType* uoff = BinaryOp::use_lhs ? (ufeat + src * ufeat_len): nullptr;
+    const DType* eoff = BinaryOp::use_rhs ? (efeat + eid * efeat_len): nullptr;
+    const DType* outoff = out + dst * out_len;
+    Idx* arguoff = BinaryOp::use_lhs ? (arg_u + dst * out_len): nullptr;
+    Idx* argeoff = BinaryOp::use_rhs ? (arg_e + dst * out_len): nullptr;
```

Comment:
And they will never be null?

Reply:
They will be null when we use `copy_u` or `copy_e` operator.